### PR TITLE
Nearest neighbor operation

### PR DESCRIPTION
**Pull request by raimis**
- [x] Move code from TorchMD-NET (https://github.com/torchmd/torchmd-net/pull/61)
- [x] Integration
- [x] Tests
- [x] Documentation

### CODE REVIEW

**Comment by raimis** [comment]
@peastman could you review?

The PBC (https://github.com/torchmd/torchmd-net/issues/92) will be add with a separate PR.

**Review by peastman — APPROVED** [review]
Looks good!  My comments at https://github.com/torchmd/torchmd-net/pull/61#issuecomment-1108854574 still apply.  For very small molecules, this should be as fast as anything.  Once you get up to several hundred atoms, there are ways it could be made faster.  Having each thread loop over multiple atoms or having multiple threads work together could reduce the number of row/column computations, position loads, and atomic adds.

File: src/pytorch/neighbors/getNeighborPairsCUDA.cu [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+template <typename scalar_t, int num_dims>
+inline Accessor<scalar_t, num_dims> get_accessor(const Tensor& tensor) {
+    return tensor.packed_accessor32<scalar_t, num_dims, RestrictPtrTraits>();
+};
+
+template <typename scalar_t> __device__ __forceinline__ scalar_t sqrt_(scalar_t x) {};
+template<> __device__ __forceinline__ float sqrt_(float x) { return ::sqrtf(x); };
+template<> __device__ __forceinline__ double sqrt_(double x) { return ::sqrt(x); };
+
+// Support pre-Kepler GPUs
```
[the remark this answers is not in the document]
Comment (raimis) [reply src/pytorch/neighbors/getNeighborPairsCUDA.cu]: Until we support CUDA 11, we have to support these cards.

File: src/pytorch/neighbors/getNeighborPairs.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        Maximum distance between atom pairs.
+    max_num_neighbors: int, optional
+        Maximum number of neighbors per atom. If set to `-1` (default),
+        all possible combinations of atom pairs are included.
+
+    Returns
+    -------
+    neighbors: `torch.Tensor`
+        Atom pair indices. The shape of the tensor is `(2, num_pairs)`.
+        If an atom pair is separated by a larger distance than the cutoff,
```
[the remark this answers is not in the document]
Comment (raimis) [reply src/pytorch/neighbors/getNeighborPairs.py:34]: Defined `num_pairs`

File: src/pytorch/neighbors/getNeighborPairs.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+from typing import Tuple
+
+
+def getNeighborPairs(positions: Tensor, cutoff: float, max_num_neighbors: int = -1) -> Tuple[Tensor, Tensor]:
+    '''
+    Returns indices and distances of atom pairs withing a given cutoff distance.
+
+    If `max_num_neighbors == -1` (default), all the atom pairs are returned,
+    (i.e. `num_atoms * (num_atoms + 1) / 2` pairs). If `max_num_neighbors > 0`,
+    only `num_atoms * max_num_neighbors` pairs are returned.
```
[the remark this answers is not in the document]
Comment (raimis) [reply src/pytorch/neighbors/getNeighborPairs.py]: I improved the text and added a few example.

**Comment by raimis** [comment]
@peastman any more comments?

**Comment by peastman** [comment]
Looks good.

**Pull request merged**